File: frontend/src/stores/monitor.ts
import { defineStore } from 'pinia'
import { ref, computed } from 'vue'
import { monitorApi } from '@/api'
import type { MonitorStatus, SchedulerStats } from '@/types'

export const useMonitorStore = defineStore('monitor', () => {
  // 状态
  const status = ref<MonitorStatus | null>(null)
  const stats = ref<SchedulerStats | null>(null)
  const loading = ref(false)
  const error = ref<string | null>(null)

  // 计算属性
  const isRunning = computed(() => status.value?.is_running ?? false)
  const totalTasks = computed(() => status.value?.total_tasks ?? 0)
  const activeTasks = computed(() => status.value?.active_tasks ?? 0)
  const runningTasks = computed(() => status.value?.running_tasks ?? 0)
  const totalChanges = computed(() => status.value?.total_changes ?? 0)
  const totalExecutions = computed(() => status.value?.total_executions ?? 0)
  const successfulExecutions = computed(() => status.value?.successful_executions ?? 0)
  const failedExecutions = computed(() => status.value?.failed_executions ?? 0)
  const uptime = computed(() => status.value?.uptime ?? 0)
  const nextCheck = computed(() => status.value?.next_check ?? null)

  // 获取监控状态
  const fetchStatus = async () => {
    try {
      error.value = null
      status.value = await monitorApi.status()
    } catch (e) {
      error.value = '获取监控状态失败'
      console.error('获取监控状态失败:', e)
    }
  }

  // 获取调度器统计
  const fetchStats = async () => {
    try {
      stats.value = await monitorApi.stats()
    } catch (e) {
      console.error('获取调度器统计失败:', e)
    }
  }

  // 刷新所有数据
  const refresh = async () => {
    await Promise.all([fetchStatus(), fetchStats()])
  }

  // 启动监控
  const startMonitor = async () => {
    loading.value = true
    error.value = null
    try {
      status.value = await monitorApi.start()
      return true
    } catch (e) {
      error.value = '启动监控失败'
      console.error('启动监控失败:', e)
      return false
    } finally {
      loading.value = false
    }
  }

  // 停止监控
  const stopMonitor = async () => {
    loading.value = true
    error.value = null
    try {
      status.value = await monitorApi.stop()
      return true
    } catch (e) {
      error.value = '停止监控失败'
      console.error('停止监控失败:', e)
      return false
    } finally {
      loading.value = false
    }
  }

  return {
    // 状态
    status,
    stats,
    loading,
    error,

    // 计算属性
    isRunning,
    totalTasks,
    activeTasks,
    runningTasks,
    totalChanges,
    totalExecutions,
    successfulExecutions,
    failedExecutions,
    uptime,
    nextCheck,

    // 方法
    fetchStatus,
    fetchStats,
    refresh,
    startMonitor,
    stopMonitor,
  }
})

File: webmon/web/api/monitor.py
"""
监控控制 API 路由

提供监控器状态查询、启动、停止的 RESTful API 端点。
"""

from fastapi import APIRouter, HTTPException, Depends

from webmon.web.schemas.monitor import (
    MonitorStatus,
    MonitorActionResponse,
    MonitorStartRequest,
    MonitorStopRequest,
    SchedulerStats,
)
from webmon.web.services.monitor_service import MonitorService, get_monitor_service

router = APIRouter(prefix="/api/monitor", tags=["monitor"])


def dict_to_monitor_status(data: dict) -> MonitorStatus:
    """将字典转换为 MonitorStatus 模型"""
    return MonitorStatus(
        status=data.get('status', 'stopped'),
        is_running=data.get('is_running', False),
        start_time=data.get('start_time'),
        uptime=data.get('uptime', 0),
        total_tasks=data.get('total_tasks', 0),
        active_tasks=data.get('active_tasks', 0),
        running_tasks=data.get('running_tasks', 0),
        total_executions=data.get('total_executions', 0),
        successful_executions=data.get('successful_executions', 0),
        failed_executions=data.get('failed_executions', 0),
        total_changes=data.get('total_changes', 0),
        next_check=data.get('next_check'),
    )


@router.get(
    "/status",
    response_model=MonitorStatus,
    summary="获取监控状态",
    description="获取监控器的当前运行状态和统计信息"
)
async def get_monitor_status(
    service: MonitorService = Depends(get_monitor_service)
) -> MonitorStatus:
    """获取监控器状态"""
    try:
        status = await service.get_status()
        return dict_to_monitor_status(status)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取监控状态失败: {str(e)}")


@router.post(
    "/start",
    response_model=MonitorActionResponse,
    summary="启动监控",
    description="启动监控器，开始执行所有启用的监控任务"
)
async def start_monitor(
    request: MonitorStartRequest = None,
    service: MonitorService = Depends(get_monitor_service)
) -> MonitorActionResponse:
    """启动监控器"""
    try:
        task_ids = request.task_ids if request else None
        result = await service.start(task_ids=task_ids)

        status = None
        if result.get('status'):
            status = dict_to_monitor_status(result['status'])

        return MonitorActionResponse(
            success=result.get('success', False),
            message=result.get('message', ''),
            status=status,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"启动监控器失败: {str(e)}")


@router.post(
    "/stop",
    response_model=MonitorActionResponse,
    summary="停止监控",
    description="停止监控器，等待所有运行中的任务完成后关闭"
)
async def stop_monitor(
    request: MonitorStopRequest = None,
    service: MonitorService = Depends(get_monitor_service)
) -> MonitorActionResponse:
    """停止监控器"""
    try:
        force = request.force if request else False
        wait_timeout = request.wait_timeout if request else 30

        result = await service.stop(force=force, wait_timeout=wait_timeout)

        status = None
        if result.get('status'):
            status = dict_to_monitor_status(result['status'])

        return MonitorActionResponse(
            success=result.get('success', False),
            message=result.get('message', ''),
            status=status,
        )
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"停止监控器失败: {str(e)}")


@router.get(
    "/stats",
    response_model=SchedulerStats,
    summary="获取调度器统计",
    description="获取调度器的详细统计信息和配置"
)
async def get_scheduler_stats(
    service: MonitorService = Depends(get_monitor_service)
) -> SchedulerStats:
    """获取调度器统计信息"""
    try:
        stats = await service.get_scheduler_stats()
        return SchedulerStats(**stats)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"获取调度器统计失败: {str(e)}")
